Mathlib/RingTheory/NoetherNormalization.lean

```diff
@@ -0,0 +1,287 @@
+      have : (T f).symm.toRingEquiv.toRingHom.comp (T f) = RingHom.id _ := by
+        apply RingEquiv.symm_toRingHom_comp_toRingHom
+      rw [this, Ideal.map_id])
+
+/-$hom2$ is the composition of maps above, from $k[X_1,...X_n]$ to $k[X_0,...X_n]/I$,
+and it's integral.-/
+private noncomputable def hom2 :=
+  (eqv2 f I).toAlgHom.comp ((eqv1 f I).toAlgHom.comp ((hom1 f I).restrictScalars k))
+
+/-Use induction to prove there is an injective map from $k[X_0,...X_{r-1}]$ to $k[X_0,...X_n]/I$.-/
+theorem exists_integral_inj (I : Ideal (MvPolynomial (Fin n) k))
+    (hi : I ≠ ⊤) : ∃ r ≤ n, ∃ g : (MvPolynomial (Fin r) k) →ₐ[k] ((MvPolynomial (Fin n) k) ⧸ I),
+    Function.Injective g ∧ g.IsIntegral := by
+  induction' n with d hd
+  · refine ⟨0, by rfl, Quotient.mkₐ k I, ?_⟩
+    constructor
+    · intro a b hab
+      rw [Quotient.mkₐ_eq_mk, Ideal.Quotient.eq] at hab
+      by_contra neq
+      apply sub_ne_zero_of_ne at neq
+      have eq := eq_C_of_isEmpty (a - b)
+      have ne : coeff 0 (a - b) ≠ 0 := fun h ↦ h ▸ eq ▸ neq <| map_zero _
+      obtain ⟨c, _, eqr⟩ := isUnit_iff_exists.mp (Ne.isUnit ne)
+      have one : c • (a - b) = 1 := by
+        rw [MvPolynomial.smul_eq_C_mul, eq, ← RingHom.map_mul, eqr, MvPolynomial.C_1]
+      exact hi ((eq_top_iff_one I).mpr (one ▸ Submodule.smul_of_tower_mem I c hab))
+    · apply isIntegral_of_surjective _ (Quotient.mkₐ_surjective k I)
+  by_cases eqi : I = 0
+  · set q := Quotient.mkₐ k I
+    have bij : Function.Bijective q := by
+      unfold q
+      simp only [Quotient.mkₐ_eq_mk]
+      constructor
+      · intro a b hab
+        rw [Ideal.Quotient.eq, eqi, Submodule.zero_eq_bot, mem_bot] at hab
+        rw [← add_zero b, ← hab, add_sub_cancel]
+      · exact Quotient.mk_surjective
+    exact ⟨d + 1, by rfl, q, bij.1, q.isIntegral_of_surjective bij.2⟩
+  · obtain ⟨f, fi, fne⟩ := Submodule.exists_mem_ne_zero_of_ne_bot eqi
+    set ϕ := kerLiftAlg <| hom2 f I
+    have ne : ker (hom2 f I) ≠ ⊤ := by
+      by_contra eq
+      replace eq := (mem_ker (f := hom2 f I) (r := 1)).mp (eq ▸ trivial)
+      simp only [AlgEquiv.toAlgHom_eq_coe, map_one] at eq
+      exact hi <| Quotient.zero_eq_one_iff.mp eq.symm
+    obtain ⟨r, _, g, injg, intg⟩ := hd (ker <| hom2 f I) ne
+    have comp : (kerLiftAlg (hom2 f I)).comp (Quotient.mkₐ k (ker (hom2 f I)))
+        = (hom2 f I) := AlgHom.ext fun a ↦ by
+      simp only [AlgHom.coe_comp, Quotient.mkₐ_eq_mk, Function.comp_apply, kerLiftAlg_mk]
+    have : (hom2 f I).IsIntegral := IsIntegral.trans _ _ (IsIntegral.trans _ _
+      (hom1_int f I fne fi) (isIntegral_of_surjective _ (eqv1 f I).surjective))
+      (isIntegral_of_surjective _ (eqv2 f I).surjective)
+    rw [← comp] at this
+    exact ⟨r, by linarith, ϕ.comp g,
+      (ϕ.coe_comp  g) ▸ (Function.Injective.comp (kerLiftAlg_injective _) injg),
+      IsIntegral.trans _ _ intg (RingHom.IsIntegral.tower_top _ _ this)⟩
+
+theorem Noether_Normalization {R : Type*} [CommRing R] [Nontrivial R] [Algebra k R]
```

Comment:
We use explicit name even for theorems that have a "standard name", so this should be called something like `exists_algHom_of_...`.